### ISSUE TITLE
Sensor lag

### ISSUE THREAD

**Issue by Firebrand** [ev3dev/ev3dev#1150]
### What are you trying to do?
Get fast and responsive readings from the sensors

### What did you expect to happen?
I expect sensor values to update within milliseconds like they do when running the Mindstorms visual environment.

### What actually happened?
I tested the code below and waved my hand in front of the UltraSonic sensor. The value would update on the screen after 2-3 secs(!). This is way too slow to actually be useful.

### What hardware and software are you using (including version numbers)?
- ev3dev-stretch-ev3-generic-2018-08-06
- Lego Mindstorms EV3

### Tested code:

```python
#!/usr/bin/env python3
from ev3dev2.display import *
from ev3dev2.fonts import *
from ev3dev2.sensor.lego import UltrasonicSensor

ultraSense = UltrasonicSensor()
ultraSense.mode = 'US-DIST-CM'
display = Display()

while 1>0:
    display.clear()
    display.draw.text((10,10), str(ultraSense.value()), font=fonts.load('luBS14'))
    display.update()
```

**Comment by dlech** [comment]
I think the problem is that you are calling `fonts.load()` inside the loop and it is re-loading the font from disk each time.

**Comment by Firebrand** [comment]
Man, thanks so much! I'm super embarrassed by my question now... getting that font load statement out of the loop should've been the first thing to try.

**Comment by Firebrand** [comment]
@dlech  Thanks for the suggestion dlech. Unfortunately, I completely removed the font loading statement and it still takes 3-4 secs to show the new value of the sensor (this happens with all sensors). This is my new code:

```python
#!/usr/bin/env python3
from ev3dev2.display import *
from ev3dev2.fonts import *
from ev3dev2.sensor.lego import UltrasonicSensor

ultraSense = UltrasonicSensor()
ultraSense.mode = 'US-DIST-CM'
display = Display()

while True:
    display.clear()
    display.draw.text((10,10), str(ultraSense.value()))
    display.update()
```

**Comment by dlech** [comment]
Do the values update faster if you view them using the Device Browser in Brickman?

**Comment by Firebrand** [comment]
@dlech Yes. If on the brick I go to Device Browser > Sensors > Watch Values on the sensor I choose, the values update in less than a second as opposed to 3-4 seconds.

**Comment by dlech** [comment]
Then there must be a problem in the python library.

cc: @ev3dev/python

**Comment by WasabiFan** [comment]
In initial testing I can confirm the issue and have isolated it to the screen drawing. Looking into this now.

**Comment by WasabiFan** [comment]
It looks like we dramatically regressed display update performance here: https://github.com/ev3dev/ev3dev-lang-python/commit/70b7b449824c996196ed5b092213b3306e7a69f2

The logic to update the display with 32bpp data [here](https://github.com/ev3dev/ev3dev-lang-python/blob/e2dd9a1180eed72e20a22f5f53744b06754b9048/ev3dev2/display.py#L310-L311) takes multiple seconds on the EV3. My inclination is to say that this is because our 32bpp image-to-byte conversion logic is implemented with Python list comprehensions whereas the Pillow equivalent (which we use for 1bpp) is [written natively in C](https://github.com/python-pillow/Pillow/blob/af552801c25002baaf3e0239206bd7bb395eaf3e/src/encode.c#L467).

**Comment by dlech** [comment]
It looks like there is [XRGB in Pack.c](https://github.com/python-pillow/Pillow/blob/81c88b1ead5343d9778a82977c9c6c1150fd9738/src/libImaging/Pack.c#L560) if you can figure out how to use that.

**Comment by dlech** [comment]
I have a feeling this might work:

    self.mmap[:] = self._img.tobytes(encoder_name='XRGB')

**Comment by WasabiFan** [comment]
Built-in encoders appear to be defined here: https://github.com/python-pillow/Pillow/blob/90a94e7dae19f75363c769713b1b6d10eb16f613/src/_imaging.c#L3694. There's no XRGB encoder (or, an encoder for any other plain pixel data format it seems) so I expect I need to look into finding an API that lets us invoke one of the "packers" that you linked to. I'll take a look today or tomorrow probably.

**Comment by WasabiFan** [comment]
I think the bottom line is that there is no existing packer defined which converts from any greyscale (single-channel) format to any color format. So while it is possible to invoke them, they will never be able to convert as we want. That being said, I think I can probably do a `merge` of the single-channel image with itself to produce a three-channel greyscale image and then operate on that.

**Comment by Firebrand** [comment]
Cool. I wish I had more Python/C exp to help you guys out with the beautiful work you do. Another ev3dev2 release comin up soon?

**Comment by WasabiFan** [comment]
I've opened ev3dev/ev3dev-lang-python#497 with my proposed fix. It gets screen drawing performance to an acceptable level, but that's still not nearly as fast as it was when we were able to directly ship the bits from the internal image buffer to the display framebuffer (as we did on ev3dev-jessie). @dlech do you have any other suggestions in this space, or is my change the best you think we are going to get?

**Comment by WasabiFan** [comment]
> Another ev3dev2 release comin up soon?

That depends on a lot of factors, but yes I do intend to make another release soon. I'll probably hold off until I can include a couple other improvements that we have queued.
